### PR TITLE
Oneshot redesign (plus ActiveModColor, et al)

### LINKED ISSUE #414: OneShot: Feature request: Add support for triggering multiple modifier OneShot
Originally by @joeynguyen, as keyboardio/Kaleidoscope-OneShot#41.

### LINKED ISSUE #416: OneShot: Possible bug: OneShot.cancel() doesn't clear the Command key (OSM(LeftGui)) on Macs
Originally by @joeynguyen, as keyboardio/Kaleidoscope-OneShot#44.

### LINKED ISSUE #419: OneShot: OneShot should support more keys
Originally by @MaxG87, as keyboardio/Kaleidoscope-OneShot#32.

### LINKED ISSUE #882: Mix LED-ActiveModColor with QuKeys
Hello,

It seems there is no way for now to make LED-ActiveModColor work on a Qukey like RALT_T(PcApplication) (PcApp is my Compose key on tap and Alt-Gr on hold), in order to let the holding Alt-Gr be colored like an Alt or a Ctrl.

Sincerely.

### LINKED ISSUE #894: Sticky OneShot keys still flagged as pressed after being released
This is a minor bug in OneShot, and something that doesn't currently result in any problems for Escape-OneShot (the only plugin I know of that calls the `OneShot.isPressed()` function), but it could easily result in problems if not addressed in the future.

When a OneShot key is double-tapped and becomes `sticky`, its `pressed` bit is also set on the second toggled-on event, but it is not cleared when the key is released. This means that `OneShot.isPressed()` will return true if there are any sticky OneShot keys, even if none are being held. Worse, it doesn't even get cleared after that key is pressed (and released) a third time to deactivate the key, so `OneShot.isPressed()` can return true even if there are no keys held, and no OneShot keys active.

If sticky OneShot keys are cancelled via Escape-OneShot, however, `OneShot.cancel(true)` does clear the `pressed` state bits, as well.

### LINKED ISSUE #895: Duplicate OneShot keys violate documented behaviour
If there are two keys currently mapped to a single OneShot `Key` value (e.g. `OSM(LeftShift)`), it is possible to tap them both in quick succession and make that OneShot `Key` sticky. Then tapping on either of the keys will cancel the OneShot. With ActiveModColor, this is easy to see. A user can even roll over (i.e. tap both simultaneously) and get the same result. Even if the two keys are pressed (almost) simultaneously and held longer than `OneShot.hold_time_out`, they will stay active (and sticky) after release.

This could be considered a feature, rather than a bug, but the documentation is fairly clear that a OneShot key should only become sticky if it is double-tapped.

### LINKED ISSUE #896: Double-tapping a OneShot layer key doesn't always make it sticky
If a OneShot layer key activates a higher layer that has a non-transparent key mapped on the OneShot key, it will stay active only until the next keypress. That works just fine, as long as the next key pressed is anything other than that OneShot key. Because Kaleidoscope does a lookup in the keymap when a key toggles on, it won't find the OneShot layer key, so it's impossible to double-tap such a key to make it sticky.

I have a solution for this problem, but it involves a major change to the interface between the event handler and the keymap, and might require changes to several plugins to adapt to it. I'll submit another issue and/or a draft PR to discuss it further. I don't think it would be worth it just to solve this rather questionable "problem", but I think it has the potential to simplify multiple keystroke-handling plugins.

### LINKED ISSUE #906: New OneShot feature: make modifiers auto-OneShot keys
OneShot could automatically turn modifier and layer-shift keys into OneShot keys. This behaviour could be enabled and disabled at runtime (e.g. with a Macros key). Additionally, it could then make combination modifiers (e.g. `Key_Meh`) into simple OneShot keys.

### LINKED ISSUE #907: New OneShot feature: meta-sticky key
The idea is to create a new special OneShot key, tentatively named `Key_MetaSticky`, which would act like a OneShot key, but would put any key on the keyboard into a OneShot "sticky" state. Possibly useful for gaming (e.g. staying crouched in Minecraft, to prevent falling off the edge of a cliff).

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
[Ideally, I'd break this apart into at least two PRs, or at least more separate commits, but I wanted to get it out here for discussion first.]

This is a major rewrite of OneShot, built on top of #904. It   also includes a corresponding update to LED-ActiveModColor, and is an alternative to #890 & #893.

## OneShot redesign

Instead of tracking the state of a set of `Key` values, this version uses two `KeyAddrBitfield` objects to track the OneShot state of each `KeyAddr` on the keyboard. This gives us two bits per key, which is enough to store four different states:
- *normal*
- *pending*
- *temporary*
- *sticky*

(I would actually prefer to use the terms "sticky" and "locked" for the last two, but that would no doubt cause too much confusion.)

### State transitions

When a *normal* OneShot `Key` toggles on, its `KeyAddr` enters the *pending* state (and the start time is set). In addition, the OneShot `Key` is temporarily replaced in the keymap cache with its corresponding modifier/layer-shift `Key` instead.

When a *pending* `KeyAddr` toggles off, it enters the *temporary* state (and stays active).

When a *temporary* `KeyAddr` toggles on, it enters the *sticky* state (unless it doesn't qualify as a double-tap, in which case it becomes *normal* instead).

When a non-modifier, non-layer-shift, non-OneShot `Key` toggles on, any `KeyAddr`s in the *pending* or *temporary* state enter the *normal* state in the next cycle (so that they will still remain active for one report).

When a *sticky* `KeyAddr` toggles on, it enters the *normal* state (and will be released when it toggles off).

### Timeouts

There hasn't been any particular change to the timeout functionality; the three usual timeouts still result in the same behaviour as before, but the underlying mechanics are different.

When a *pending* `KeyAddr`'s hold timeout expires, it enters the *normal* state.

When a *temporary* `KeyAddr`'s timeout expires, it enters the *normal* state if it's still held, or gets released otherwise.

If a *temporary* `KeyAddr` toggles on, if the double-tap timeout has expired (or it wasn't the immediately-previous `KeyAddr` to toggle on), it enters the *normal* state instead of becoming *sticky*.

### Keymap cache interactions

The new version of OneShot works by immediately decoding OneShot keys when they're pressed, and putting the corresponding "normal" key into the keymap cache by calling `handleKeyswitchEvent()` with a `Key` value, then aborting the original event by returning `EVENT_CONSUMED`.

Later, when the *temporary* `KeyAddr` toggles off, that release event is likewise aborted, stopping `handleKeyswitchEvent()` from clearing its entry in the keymap cache.

By doing it this way, we make it possible for a OneShot layer-shift key to remain in effect when it is double-tapped to make it sticky, even if the target layer has a different `Key` mapped to the OneShot's `KeyAddr`. This could be viewed as a disadvantage (vs current behaviour), because whatever that `Key` was is not accessible via the OneShot layer-shift. It is more consistent behaviour, however, so I think it's an improvement.

### Automatic one-shot modifiers

OneShot is now capable of automatically treating modifiers (and/or layer-shift keys) as one-shot keys. With auto-modifiers enabled, it is possible to turn a modifier with additional modifier flags into a single one-shot key. While `OSM(Key_Meh)` doesn't work, `enableAutoModifiers()` will allow `Key_Meh` on its own to become a one-shot key.

Also noteworthy; auto-modifiers only applies to keys with keycode byte that is one of the eight modifiers, so it's possible to have a modifier key in a keymap that doesn't behave like a one-shot key even when it's turned on, by mapping a key to `LSHIFT(Key_NoKey)`, for example.

Auto one-shot modifiers and layer-shifts can be turned on and off independently or together. A Macros key can be used to toggle either or both.

### `Key_MetaSticky`

I added a special OneShot key, tentatively named `Key_MetaSticky`, which behaves like a OneShot key when pressed, and makes the next key pressed sticky, regardless of its value. This has very limited utility, but could certainly come in handy for some games that use the keyboard keys for something other than text.

## LED-ActiveModColor

I used a similar technique here, with a `KeyAddrBitfield` to keep track of `KeyAddr`s representing active modifier and layer-shift `Key` values. Furthermore, it is now able to used three different colors to hightlight keys of three different categories:

- *normal* modifier/layer-shift keys that become inactive when released (white)
- *temporary* modifier/layer-shift keys that will stay active after release (yellow)
- *sticky* modifier/layer-shift keys that will stay active indefinitely (red)

In my limited experience, it's nice to have a visual indicator that a key's OneShot hold timeout has expired. It's not a huge feature, but I think it's worthwhile.

## Escape-OneShot

Escape-OneShot is simplified; it no longer needs to track anything in order to suppress the held Escape key, because it can simply call `Layer.updateLiveCompositeKeymap()` to mask the `KeyAddr` temporarily by setting it to `Key_NoKey`.

I have also added a way to configure which `Key` value will be used to cancel one-shot (and sticky) keys, and a special `Key_OneShotEscape` key whose sole purpose is to have a key that can be used to dedicate a key to cancelling one-shot keys without any side effects (with the default configuration, the user would need to use a modifier chorded in order to get a modified <kbd>esc</kbd> key).

---
## Issues addressed

Fixes #882
Fixes #894
Fixes #895
Fixes #896
Closes #906 
Closes #907
Closes #419
Closes #414 (debatably)
Closes #416 (debatably)
 
## Supersedes other PRs

Closes #890 
Closes #893

### CODE REVIEW

**Comment by gedankenexperimenter** [comment]
The one possible showstopper I foresee is that I don't have a way to make OneShot keys without a valid `KeyAddr` work. So, for example, if a Macro sends a OneShot, it wouldn't work without specifying a key address somehow. A brief look at Macros indicates that it marks all of its events as `INJECTED`, however, which means OneShot should already be ignoring them, so that particular example doesn't represent a loss of functionality.

**Comment by gedankenexperimenter** [comment]
Something else that can be done with this version: I could trivially add configurable functionality that makes modifiers and/or layer-shift keys automatic OneShot keys. This could apply to not only simple keycode-only Keyboard  modifiers, but also (optionally) ones with modifier flags, making single-key modifier combos possible without the use of Macros.

**Comment by gedankenexperimenter** [comment]
I went ahead and added the auto-OneShot modifiers and/or layer-shift keys feature.

**Comment by gedankenexperimenter** [comment]
Next, I'll add a special OneShot key that acts as a modifier to make any key a OneShot key.

**Comment by gedankenexperimenter** [comment]
New feature added: `Key_MetaSticky` (I'm happy to rename it, of course—that's just what came to mind). This is a key that becomes sticky immediately when pressed, and while it remains active, makes other keys sticky, regardless of their value. All those keys (and any other active OneShot keys) get released when `Key_MetaSticky` is pressed and released a second time.

I figured this would be more useful than turning normal keys into OneShot keys, but might come in handy in certain circumstances. It would probably be more useful to have the other sticky keys remain active after `Key_MetaSticky` gets release, now that I think about it…

**Comment by gedankenexperimenter** [comment]
I improved `Key_MetaSticky`, which now behaves like a OneShot key, rather than being auto-sticky. It makes other keys _sticky_, and does not clear their _sticky_ state when it becomes inactive itself.

**Comment by algernon** [comment]
Just a short note: I've looked into this PR and #904 yesterday, and so far, I'm liking what I'm seeing. Will come back to them when I'm done with other priority tasks.

**Comment by gedankenexperimenter** [comment]
Thanks for taking a look, @algernon. I'm not entirely happy with this one yet. In particular, compared to #893, This one results in a significantly larger binary, especially with the `Key_MetaSticky` feature, which is more complex than I anticipated. I'm thinking of trying to separate that from the rest of the code somehow, so that users who don't want it don't have to waste a couple hundred bytes of PROGMEM. It might work to make it a second plugin ("OneShot-MetaSticky"?), but I'm not sure.

Also, with `Key_MetaSticky`, I'm currently getting an extra toggle-off and toggle-on when I press a letter key for the second time to cancel it being sticky, and I haven't figured out why yet. It's not something that's likely to get noticed, but it's still a bug.

**Comment by gedankenexperimenter** [comment]
I should also mention that the bitfield design and the auto-OneShot and "meta-sticky" features are based on code I wrote for Kaleidoglyph-Glukeys a while ago.

**Comment by gedankenexperimenter** [comment]
I fixed the meta-sticky bug that caused an extra toggle-off & toggle-on when cancelling a stick letter key, and added a bunch of `#ifdef`s to conditionally compile without the `Key_MetaSticky` feature and/or without deprecated stuff.

**Comment by gedankenexperimenter** [comment]
Speaking of deprecated stuff, I'm playing around with a trick for deprecating member variables that were public, but we might want to be private. It's not critical in this case, since there's no value the user can use that will break anything, but I think it makes sense to move in that direction, and I saw a way to do it.

**Comment by gedankenexperimenter** [comment]
It looks like TravisCI is having trouble building right now.

**Comment by gedankenexperimenter** [comment]
I edited the summary comment at the top to keep it updated with all the changes I've made. I think I'm done adding features at this point, but if I add anything else, I'll update it again.

**Comment by gedankenexperimenter** [comment]
This should get rebased onto `master`, but I'd rather not do it more than once, so I'll wait until we can have a discussion about it (and #904) before doing any of that work.

**Comment by gedankenexperimenter** [comment]
I've been looking at old OneShot feature requests and issues. A few of them are addressed by this PR. On the downside, it makes #424 very difficult (almost anathema).

**Comment by gedankenexperimenter** [comment]
So, this is now failing the cpplint check because of the conditional compilation I put in for leaving out the `Key_MetaSticky` feature. I'm not sure how to make this pass and still let some people compile it without a costly feature that they don't want. I've already inserted comments to pacify astyle, which would otherwise mess up the indentation. I'll see what I can do, but I'm having trouble finding documentation for cpplint.

**Comment by gedankenexperimenter** [comment]
Also, when I run `make cpplint` locally, I get 138 errors, from many files. I guess Travis-CI is using different settings? Is there a documented way for me to reproduce what Travis is doing before committing changes?

**Comment by gedankenexperimenter** [comment]
One other big advantage to this version over the current master: if you double-tap to make a layer-shift sticky with this version, it's impossible to get stuck in the target layer, regardless of the keymap in that layer, because the OSL key will still be active as on OSL, overriding whatever that key is mapped to on the target layer, so tapping it again will cancel it. See [this conversation on the forum](https://community.keyboard.io/t/help-me-set-up-layer-switching-how-i-want/4214) for an example use-case.

**Comment by gedankenexperimenter** [comment]
I've got a bit more work to do on this before it should be merged. I think the code is done, but I want to add removal dates for deprecated stuff and an update to UPGRADING.md. I'd really like to add at least a few testcases, too, but since we're just getting started with that, I'm feeling that is not as essential right now. In the future, I'd like to included testcases with every PR, and in the case of bugfixes, start those PRs with a commit of a failing testcase, if possible.

**Comment by gedankenexperimenter** [comment]
One more thing to add: A special key like `Key_MetaSticky` that makes any currently held key sticky when it toggles on. In fact, that's probably more useful and easier to implement than what I've already done. See [this thread in the forum](https://community.keyboard.io/t/tweak-qukeys-with-additional-keypress-to-perform-sticky-locklayer-n/4267/3).

**Comment by gedankenexperimenter** [comment]
I added the code for one last feature. A couple of names need to be changed, and the commits reorganized, and docs updated. And, most important, tests written. I'm going to lobby for the PR to get reviewed before an exhaustive set of tests are written, as I know of users who would like to have access to some of the features and bugfixes that this PR provides. I hope to be able to mark it ready for review tomorrow, after which I'll start working on some tests, which I would like to take my time with so that they'll be of fairly high quality.

**Comment by gedankenexperimenter** [comment]
> These debug statements may not want to get in here (without at least a debugging ifdef)

Oops! Yeah, this change is definitely not done. I pushed it because someone is testing it out, and needed a bugfix, and I'm pretty sure I got it working. It was late, and the PR is still a draft, so I figured it was best to push it anyway, but leaving in the debug statements was clearly an oversight!

**Comment by gedankenexperimenter** [comment]
I just made a big update to the commit history. This still needs some documentation updates, and testcases.

File: src/kaleidoscope/plugin/OneShot.cpp [hunk 1]
```diff
@@ -18,270 +18,526 @@
...
+
+    // Make all held keys sticky if `Key_OneShot_Stickify` toggles on.
+    if (key == Key_OneShot_Stickify) {
+      // Skip the stickify key itself
+      for (KeyAddr entry_addr : KeyAddr::all()) {
+        if (entry_addr == key_addr) {
+          continue;
         }
+        // Get the entry from the keymap cache
+        Key entry_key = Layer.lookup(entry_addr);
```
Comment (gedankenexperimenter) [inline src/kaleidoscope/plugin/OneShot.cpp:216]: This needs to change to `Runtime.activeKey(entry_addr)`.

**Comment by gedankenexperimenter** [comment]
I decided to try renaming the two new special OneShot "sticky-helper" keys as class constants:
```c++
OneShot::PrepStickyKey
OneShot::HeldStickyKey
```
…but that causes linker errors, so it's not an option. I'm still struggling to come up with names that aren't unweildy. The best I've got so far is:
```c++
OneShot_PreStickyKey
OneShot_PostStickyKey
```
That at lest hints at how to use them. The "pre-sticky" one gets pressed before the key that becomes sticky, the "post-sticky" one gets pressed after—but it also needs that key to still be held, so it's not such a great name. What it really does is make all _active_ keys sticky, so if you tap a few OneShot keys, and they're still active, then hit the "post-sticky" key, those OneShots will become sticky.

Maybe it would be better to drop the "sticky" label, even though that's what OneShot already uses. As I've said before "sticky" seems to me like the one-shot state, and what OneShot calls "sticky" I would call "locked". Maybe "lock" will work better for these keys, because it's not an adjective? What about `OneShot_LockActiveKey` or something like that? But then I'm not sure what to call the other one. Maybe the "meta" label is actually still good for that one, since it does behave like a OneShot key, but its effect is to make other keys sticky.

**Comment by gedankenexperimenter** [comment]
The names I've got now are `OneShot_MetaStickyKey` & `OneShot_ActiveStickyKey`, because I think those are the most descriptive without being excessively wordy (though they are debatably already excessive). They're defined in the global namespace in order to make it convenient to put those values in the keymap. I've got an idea for improving that, but it's basically unrelated to this PR.

**Comment by jeancroy** [comment]
Does this play well with QuKey ?

Current OSM don't really work with dual use key.
For example: press OSM shift, press or hold Dual use "e" / Layer Shift

Expected behavior is to either consume modifier together with "e"
Or keep modifier available on next layer shift.

Current behavior is cancel OSM now, and at a latter point send either send a plain "e" or a layer shift... 
this is terrible.

**Comment by gedankenexperimenter** [comment]
> Current master behavior is cancel OSM now, and at a latter point send either send a plain "e" or a layer shift...
> this is terrible.

In what order are the two plugins initialized in your sketch? This is what I would expect if the OneShot event handler is called before the Qukeys one, but not vice versa. It's possible, even somewhat likely, that this PR would improve certain interactions between OneShot and Qukeys, but I don't expect the behaviour you described with master as long as Qukeys comes before OneShot (and, in fact, most other plugins) in `KALEIDOSCOPE_INIT_PLUGINS()`.

**Comment by jeancroy** [comment]
Thanks. I'm using dygma raise.

You suggest Qukeys should go sooner in that list ? Is this the rigth file?
https://github.com/Dygmalab/Raise-Firmware/blob/017da140d81b975afc14dcc0aa8aac5e92db5442/Raise-Firmware.ino#L207

Sorry this is becoming off topic to this PR.

**Comment by gedankenexperimenter** [comment]
> You suggest Qukeys should go sooner in that list ? Is this the rigth file?
> https://github.com/Dygmalab/Raise-Firmware/blob/017da140d81b975afc14dcc0aa8aac5e92db5442/Raise-Firmware.ino#L207

Yes. It should be moved up above TapDance, at least.

**Pull request closed without merging**